File: backend/app/schemas/ui_healing.py
"""
UI 自动化自愈 Schema
"""
from typing import Optional, List, Any
from datetime import datetime
from pydantic import BaseModel, Field


# ==================== 页面访问记录 ====================

class PageVisitResponse(BaseModel):
    id: int
    project_id: int
    script_id: Optional[int] = None
    run_id: Optional[int] = None
    page_url: str = ""
    page_title: str = ""
    page_identifier: str = ""
    action_type: str = ""
    target_selector: str = ""
    action_result: str = ""
    fail_reason: str = ""
    source: str = "execution"
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PageVisitListResponse(BaseModel):
    total: int
    items: List[PageVisitResponse]


# ==================== 页面画像 ====================

class PageProfileResponse(BaseModel):
    id: int
    project_id: int
    page_identifier: str
    page_name: str = ""
    page_description: str = ""
    key_elements: Optional[Any] = None
    success_paths: Optional[Any] = None
    failure_patterns: Optional[Any] = None
    reachable_from: Optional[Any] = None
    visit_count: int = 0
    success_rate: float = 1.0
    last_aggregated_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PageProfileListResponse(BaseModel):
    total: int
    items: List[PageProfileResponse]


# ==================== 元素指纹 ====================

class ElementFingerprintResponse(BaseModel):
    id: int
    project_id: int
    page_identifier: str = ""
    element_role: str = ""
    element_text: str = ""
    selectors: Optional[Any] = None
    attributes: Optional[Any] = None
    occurrence_count: int = 1
    success_count: int = 0
    fail_count: int = 0
    is_stable: bool = False
    last_seen_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ==================== 自愈记录 ====================

class HealingRecordResponse(BaseModel):
    id: int
    project_id: int
    script_id: Optional[int] = None
    run_id: Optional[int] = None
    step_index: Optional[int] = None
    page_url: str = ""
    page_identifier: str = ""
    original_selector: str = ""
    action_type: str = ""
    fail_reason: str = ""
    healing_level: str = ""
    healing_strategy: str = ""
    suggested_selector: str = ""
    ai_reasoning: str = ""
    candidates: Optional[Any] = None
    healing_result: str = ""
    applied_to_script: bool = False
    confirmed_by: Optional[int] = None
    confirmed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class HealingRecordListResponse(BaseModel):
    total: int
    items: List[HealingRecordResponse]


class HealingConfirmRequest(BaseModel):
    apply_to_script: bool = Field(True, description="是否同时回写到脚本")


class HealingStatsResponse(BaseModel):
    total: int = 0
    success: int = 0
    failed: int = 0
    pending_review: int = 0
    l1_count: int = 0
    l2_count: int = 0
    l3_count: int = 0
    l4_count: int = 0
    applied_count: int = 0
    success_rate: float = 0.0

File: frontend/src/api/uiHealing.ts
import request from './request'

// ==================== 自愈记录 ====================

export interface HealingRecord {
  id: number
  project_id: number
  script_id?: number
  run_id?: number
  step_index?: number
  page_url: string
  page_identifier: string
  original_selector: string
  action_type: string
  fail_reason: string
  healing_level: string
  healing_strategy: string
  suggested_selector: string
  ai_reasoning: string
  candidates?: any
  healing_result: string
  applied_to_script: boolean
  confirmed_by?: number
  confirmed_at?: string
  screenshot_before?: string
  screenshot_after?: string
  created_at?: string
}

export interface HealingStats {
  total: number
  success: number
  failed: number
  pending_review: number
  l1_count: number
  l2_count: number
  l3_count: number
  l4_count: number
  applied_count: number
  success_rate: number
}

export function listHealingRecords(params: {
  project_id: number
  script_id?: number
  run_id?: number
  healing_level?: string
  healing_result?: string
  page?: number
  page_size?: number
}) {
  return request.get('/ui-healing/records', { params })
}

export function getHealingRecord(id: number) {
  return request.get(`/ui-healing/records/${id}`)
}

export function confirmHealing(id: number, apply_to_script = true) {
  return request.post(`/ui-healing/records/${id}/confirm`, { apply_to_script })
}

export function getHealingStats(project_id: number): Promise<HealingStats> {
  return request.get('/ui-healing/stats', { params: { project_id } })
}

// ==================== 页面画像 ====================

export interface PageProfile {
  id: number
  project_id: number
  page_identifier: string
  page_name: string
  page_description: string
  key_elements?: any
  success_paths?: any
  failure_patterns?: any
  reachable_from?: any
  visit_count: number
  success_rate: number
  last_aggregated_at?: string
  created_at?: string
  updated_at?: string
}

export function listPageProfiles(params: {
  project_id: number
  keyword?: string
  page?: number
  page_size?: number
}) {
  return request.get('/ui-healing/page-profiles', { params })
}

export function getPageProfile(id: number): Promise<PageProfile> {
  return request.get(`/ui-healing/page-profiles/${id}`)
}

// ==================== 元素指纹 ====================

export function listElementFingerprints(params: {
  project_id: number
  page_identifier?: string
  keyword?: string
  page?: number
  page_size?: number
}) {
  return request.get('/ui-healing/element-fingerprints', { params })
}

// ==================== 手动聚合 ====================

export function triggerAggregation(project_id: number) {
  return request.post('/ui-healing/aggregate', null, { params: { project_id } })
}
